crates/matrix-sdk-base/src/sliding_sync.rs

```diff
@@ -363,6 +368,49 @@ impl BaseClient {
     }
 }
 
+/// Find the most recent decrypted event and cache it in the supplied RoomInfo.
+/// If any encrypted events are found after that one, store them in the RoomInfo
+/// too so we can use them when we get the relevant keys.
+fn cache_latest_events(room_info: &mut RoomInfo, events: &[SyncTimelineEvent]) {
+    let mut encrypted_events = Vec::with_capacity(room_info.latest_encrypted_events.capacity());
+    for e in events.iter().rev() {
+        let timeline_event: serde_json::Result<AnySyncTimelineEvent> = e.event.deserialize();
+        if let Ok(timeline_event) = timeline_event {
+            match timeline_event.event_type() {
+                TimelineEventType::RoomEncrypted => {
+                    // m.room.encrypted - this might be the latest event later - we can't tell until
+                    // we are able to decrypt it, so store it for now
+                    //
+                    // Check how many encrypted events we have seen. Only store another if we
+                    // haven't already stored the maximum number.
+                    if encrypted_events.len() < encrypted_events.capacity() {
+                        encrypted_events.push(e.event.clone());
+                    }
+                }
+                TimelineEventType::RoomMessage => {
+                    // m.room.message - we found one! Store it.
+                    room_info.latest_event = Some(e.clone());
+                    // We don't need any of the older encrypted events because we have a new
+                    // decrypted one.
+                    room_info.latest_encrypted_events.clear();
+                    // We can stop looking through the timeline now because everything else is
+                    // older.
+                    break;
+                }
+                _ => {
+                    // Ignore other event types
+                }
+            }
+        } else {
+            timeline_event.unwrap();
```

Comment:
Please no unwraps, if we're sure that the unwrap won't panic use `except()` instead.

Reply:
After my tidying and finishing off, the remaining unwraps are in tests, except for:

`self.inner.read().unwrap()` and `self.inner.write().unwrap()` in `Room`, which follow the pattern we were already using in that code, so I think they are OK.